responders/Gmail/Gmail.py

```diff
@@ -0,0 +1,224 @@
+                quote(self.get_param("config.gmail_client_email", None, "Client email missing"))
+            )
+        }
+
+        self.__gmail_service = None
+        self.__hive_service = None
+
+    def __not_found(self):
+        self.error("service named {} not found.".format(self.service))
+
+    def __get_gmail_subjects(self, caseId, query):
+        """
+        Get all email addresses of a case for the correspondig gsuite domain
+
+        Returns: Array of Observable objects on success
+        """
+        response =  self.__hive_service.get_case_observables(caseId, query=query)
+        if response.status_code == 200:
+            return response.json()
+        else:
+            self.error("Failed to get valid response for query: {}".format(response.status_code, response.text))
+
+    def __get_filter_tag(self, tags):
+        """
+        Get the correct tag for a dataType in a list of tags
+
+        Returns: tag string on success else None
+        """
+        for tag in tags:
+            if "gmail_filter:{}".format(self.get_param("data.data")) in tag:
+                return tag
+        self.error("No valid filter tag found on observable. Tags: {}".format(tags))
+
+    def hive_auth(self, url, api_key):
+        self.__hive_service = TheHiveApi(url, api_key)
+        try:
+            self.__hive_service.health()
+        except TheHiveException as e:
+            self.error("Responder needs TheHive connection but failed: {}".format(e))
+
+    def gmail_impersonate(self, subject):
+        """Peforms OAuth2 auth for a given service account, scope and a delegated subject
+
+        Args:
+            subject (str): email adress of the user, whos data shall be accessed (delegation)
+
+        Returns:
+            google.auth.service_account.Credentials if valid otherwise None
+        """
+        credentials = service_account.Credentials.from_service_account_info(
+            info=self.__gmail_service_account,
+            scopes=self.__scopes,
+            subject=subject
+        )
+
+        if credentials.has_scopes(self.__scopes):
+            return build("gmail", "v1", credentials=credentials)
+        else:
+            self.error("Gmail service account creation failed. Aborting responder")
+
+    def trash_message(self, case_id, query):
+        """Moves specified message into trash. this emails can be recovered if false-positive
+        """
+        gmail_observables = self.__get_gmail_subjects(case_id, And(Eq("dataType", "mail"), EndsWith("data", self.__gmail_domain)))
+        for observable in gmail_observables:
+            resource = self.gmail_impersonate(observable["data"])
+            try:
+                response = resource.users().messages().list(userId=observable["data"], q=query).execute()
+                for message in response.get("messages", []):
+                    result = resource.users().messages().trash(userId=observable["data"], id=message["id"]).execute()
```

Comment:
Hello @strassi! 
Great job with the responder!
What do you think about adding `.delete` function as well? 
There are some cases when the email's nature is so dangerous, so moving to trash is not an option (as the users still could read it if they go and check the trash), only complete deleting could be used.

Reply:
> What do you think about adding .delete function as well?

I thought about adding a delete service as well, but did not implement it because of these reasons:

1. The difference between the actions `.trash` and `.delete` might not be clear at first. A False-Positive could lead to data loss if `.delete` was used.
2. Deletion is done by running a Gmail search query against the users mailbox. A faulty query could erase an entire mailbox.

In my opinion, the `.trash`ing of messages is a safe default that enables easy recovery. I also might argue, that a user does not dig up random trashed messages. In my experience they search for specific trashed messages in case of recovery.

 > There are some cases when the email's nature is so dangerous, so moving to trash is not an option (as the users still could read it if they go and check the trash), only complete deleting could be used.

Can you provide an example for such a message?

Reply:
I see your point, and it totally makes sense. This topic is quite complex and has a few not obvious things to consider. Let me highlight some of them:

> The difference between the actions `.trash` and `.delete` might not be clear at first. A False-Positive could lead to data loss if `.delete` was used.

Removing the email is a Response Action from the Eradication stage:

![image](https://user-images.githubusercontent.com/13090109/98456152-3c84e080-217a-11eb-8c2f-8aa8d99d2251.png)

This stage is harmful by definition — it's about deletion files, mails, services, and any kind of threat actor presence.
At the same time, it's not a stage when you can change your mind and re-classify an Incident to a False Positive. 
That a task for the Identification stage:

![image](https://user-images.githubusercontent.com/13090109/98456180-a1403b00-217a-11eb-9cb8-f230f1448d8d.png)

The Identification stage is also a stage when you collect the evidence, make a copy of it, and put it in a safe place for further analysis/processing/legal procedures. 
Getting to the Eradication stage not having a copy of evidence is not the best way to handle the incident. 

> Deletion is done by running a Gmail search query against the users mailbox. A faulty query could erase an entire mailbox.

I believe, the nature of the Eradication stage could be extremely harmful in the wrong hands. But IR people supposed to know what they are doing, configuring firewall rules, deleting files, changing configuration, etc. I believe that it is a normal thing, that some Responder during the Eradication *could* cause issues like the one you've described. But I don't think that this is somehow connected with your Responder, as this is a tool. It should be powerful. If somebody erased user's mailbox entirely, it's totally not the Cortex Responder issues. It's the Incident Response specialist issue.

> Can you provide an example for such a message?

Yep. There were many cases with malicious macros in MS Office files, attached to the emails.
User opens it, does a few clicks — boom, the second stage is downloaded, executed, and everything is on fire. 

The most horrible part of it is that some users in the organization "A" (which we were monitoring) were sending this email to their friends (asking "hey do you know why it doesn't work? Could you please try to open it for me?" etc) in the organization "B", that fortunately was also under our monitoring. 
So one incident related to one organization became a multi-organization incident. 

We also need to keep in mind that threat actors can ask victims to check their trash/spam mailboxes using other channels (LinkedIn, instant messengers, other social media). 
That's why in some cases we have to completely remove the malicious emails.

---

I believe that having both `.trash`ing and `.delete`ing options is the best way (: 
We can develop special notification about `.delete` function, providing all the details and warnings to the users of your Responder, this way, providing people with the freedom to choose. 

What do you think, @strassi?

Reply:
I thought about your arguments concerning the IR eradication stage. After reassessing my thoughts about the `.delete` API call I came to following conclusion.

I would swap the `.trash` with the `.delete` API.
This way the action cannot be misunderstood and it serves the single purpose of cleaning out a mailbox as a IR specialist would expect. A second advantage of simply swapping the API calls would be: Less code, less errors.

Unfortunately I cannot delete messages via the [filter settings API](https://developers.google.com/gmail/api/guides/filter_settings). Filters can only move the message to trash at best. 

Would you see this is as a possible solution?

Reply:
I've found one, I am not sure if that's the best of possible options, but it works. 

The workflow is the following:

1. Grab all users in the domain using Admin API
2. Execute filter query per each account 
3. Delete email by message ID if found

How to make it work:

1. Enable [Admin API](https://developers.google.com/admin-sdk)
2. Add the `https://www.googleapis.com/auth/admin.directory.user.readonly` to the existing service account scopes
3. Request new `credentials.json` (I am not sure if this step is required, did it just in case as the new permissions could not be applied to the old credentials. At least that's how it works for other APIs I've worked with)
4. Use the code snippet below to delete messages using filter query:

```python
#!/usr/bin/env python3

from google.oauth2 import service_account
from googleapiclient.discovery import build

SUBJECT = "impersonation@domain.com"

GMAIL_SCOPES = [ 
    "https://mail.google.com/", 
    "https://www.googleapis.com/auth/gmail.settings.basic",
]

ADMIN_SCOPES = [ 
    "https://www.googleapis.com/auth/admin.directory.user.readonly"
]

SERVICE_ACCOUNT_FILE = "/path/to/credentials.json"

TARGET_DOMAIN='domain.com'  # Service account will access data from this domain.

#
# Fill in a list of user IDs (emails)
#

credentials = service_account.Credentials.from_service_account_file(
        SERVICE_ACCOUNT_FILE, subject=SUBJECT, scopes=ADMIN_SCOPES)

admin_service = build("admin", "directory_v1", credentials=credentials)
results = admin_service.users().list(domain=TARGET_DOMAIN).execute()
users = results.get('users', [])


all_users = []
for user in users:
    all_users.append(user['primaryEmail'])

#
# Remove emails that match our query for each user
#

query = "from:badguy@malicious.com"

for user_id in all_users:
    credentials = service_account.Credentials.from_service_account_file(
        SERVICE_ACCOUNT_FILE, subject=user_id, scopes=GMAIL_SCOPES)
    gmail_service = build("gmail", "v1", credentials=credentials)
    response = gmail_service.users().messages().list(userId=user_id,q=query).execute()
    if response.get('messages'):
        for message in response['messages']:
            gmail_service.users().messages().delete(userId=user_id, id=message['id']).execute()

```

Let me know what do you think about it (:

Reply:
In my opinion the responder should not be run over the entire user base on its own. The observables in a case serve the purpose of documenting IOCs and affected objects/subjects. At least this is what I use them for.

I think it is quite a good practice to isolate the work of a responder does as much as possible. Too much logic takes away the control of the incident responder.

I would propose changing the `.trash` call to `.delete`, but leave the filter behavior as it is for now. If a IR specialist demands a "erase" functionality over the whole user domain we can add it in later via a new pull request. 

You could think of this responder with the current feature set as a MVP.

Reply:
If you would like to finish at this point, that's totally fine (: You did a great job, developing the Responder.

I will create an issue and provide my snippet that could help to develop the following functionality:

- [RA4201](https://atc-project.github.io/atc-react/Response_Actions/RA_4201_delete_email_message/): Delete email message from all the domain's users
- [RA2201](https://atc-project.github.io/atc-react/Response_Actions/RA_2201_list_users_opened_email_message/): List users opened email message
- [RA2203](https://atc-project.github.io/atc-react/Response_Actions/RA_2203_list_email_message_receivers/): List email message receivers

I think that your responder will have a great future and a lot of users will benefit from it!

Reply:
Yes, I would propose to finish the responder.

I'm going to swap the `.trash` with the `.delete` api call.